File: java/com/epicdragonworld/gameserver/network/SendablePacket.java
package com.epicdragonworld.gameserver.network;

import java.io.ByteArrayOutputStream;

/**
 * @author Pantelis Andrianakis
 */
public class SendablePacket
{
	private final ByteArrayOutputStream _baos;
	
	public SendablePacket()
	{
		_baos = new ByteArrayOutputStream();
	}
	
	public void writeString(String value)
	{
		if (value != null)
		{
			try
			{
				final byte[] byteArray = value.getBytes("UTF-8");
				writeByte(byteArray.length);
				writeBytes(byteArray);
			}
			catch (Exception e)
			{
				_baos.write(0);
			}
		}
		else
		{
			_baos.write(0);
		}
	}
	
	public void writeBytes(byte[] array)
	{
		try
		{
			_baos.write(array);
		}
		catch (Exception e)
		{
		}
	}
	
	public void writeByte(int value)
	{
		_baos.write(value & 0xff);
	}
	
	public void writeShort(int value)
	{
		_baos.write(value & 0xff);
		_baos.write((value >> 8) & 0xff);
	}
	
	public void writeInt(int value)
	{
		_baos.write(value & 0xff);
		_baos.write((value >> 8) & 0xff);
		_baos.write((value >> 16) & 0xff);
		_baos.write((value >> 24) & 0xff);
	}
	
	public void writeLong(long value)
	{
		_baos.write((int) (value & 0xff));
		_baos.write((int) ((value >> 8) & 0xff));
		_baos.write((int) ((value >> 16) & 0xff));
		_baos.write((int) ((value >> 24) & 0xff));
		_baos.write((int) ((value >> 32) & 0xff));
		_baos.write((int) ((value >> 40) & 0xff));
		_baos.write((int) ((value >> 48) & 0xff));
		_baos.write((int) ((value >> 56) & 0xff));
	}
	
	public void writeFloat(float fvalue)
	{
		final int value = Float.floatToRawIntBits(fvalue);
		_baos.write(value & 0xff);
		_baos.write((value >> 8) & 0xff);
		_baos.write((value >> 16) & 0xff);
		_baos.write((value >> 24) & 0xff);
	}
	
	public void writeDouble(double dvalue)
	{
		final long value = Double.doubleToRawLongBits(dvalue);
		_baos.write((int) (value & 0xff));
		_baos.write((int) ((value >> 8) & 0xff));
		_baos.write((int) ((value >> 16) & 0xff));
		_baos.write((int) ((value >> 24) & 0xff));
		_baos.write((int) ((value >> 32) & 0xff));
		_baos.write((int) ((value >> 40) & 0xff));
		_baos.write((int) ((value >> 48) & 0xff));
		_baos.write((int) ((value >> 56) & 0xff));
	}
	
	public byte[] getSendableBytes()
	{
		// Encrypt bytes.
		final byte[] encryptedBytes = Encryption.encrypt(_baos.toByteArray());
		final int size = encryptedBytes.length;
		
		// Create two bytes for length (short - max length 32767).
		final byte[] lengthBytes = new byte[2];
		lengthBytes[0] = (byte) (size & 0xff);
		lengthBytes[1] = (byte) ((size >> 8) & 0xff);
		
		// Join bytes.
		byte[] result = new byte[size + 2];
		System.arraycopy(lengthBytes, 0, result, 0, 2);
		System.arraycopy(encryptedBytes, 0, result, 2, size);
		
		// Return the data.
		return result;
	}
}

File: EpicDragonWorldServer/Network/SendablePacket.cs
using System;
using System.IO;
using System.Text;

/**
 * Author: Pantelis Andrianakis
 * Date: December 23rd 2017
 */
public class SendablePacket
{
    private readonly MemoryStream memoryStream;

    public SendablePacket()
    {
        memoryStream = new MemoryStream();
    }

    public void WriteString(string value)
    {
        if (value != null)
        {
            byte[] byteArray = Encoding.UTF8.GetBytes(value);
            // Since we use short value maximum byte size for strings is 32767.
            // Take care that maximum packet size data is 32767 bytes as well.
            // Sending a 32767 byte string would require all the available packet size.
            WriteShort(byteArray.Length);
            WriteBytes(byteArray);
        }
        else
        {
            memoryStream.WriteByte(0);
        }
    }

    public void WriteBytes(byte[] byteArray)
    {
        for (int i = 0; i < byteArray.Length; i++)
        {
            memoryStream.WriteByte(byteArray[i]);
        }
    }

    public void WriteByte(int value)
    {
        memoryStream.WriteByte((byte)value);
    }

    public void WriteShort(int value)
    {
        memoryStream.WriteByte((byte)value);
        memoryStream.WriteByte((byte)(value >> 8));
    }

    public void WriteInt(int value)
    {
        memoryStream.WriteByte((byte)value);
        memoryStream.WriteByte((byte)(value >> 8));
        memoryStream.WriteByte((byte)(value >> 16));
        memoryStream.WriteByte((byte)(value >> 24));
    }

    public void WriteLong(long value)
    {
        memoryStream.WriteByte((byte)value);
        memoryStream.WriteByte((byte)(value >> 8));
        memoryStream.WriteByte((byte)(value >> 16));
        memoryStream.WriteByte((byte)(value >> 24));
        memoryStream.WriteByte((byte)(value >> 32));
        memoryStream.WriteByte((byte)(value >> 40));
        memoryStream.WriteByte((byte)(value >> 48));
        memoryStream.WriteByte((byte)(value >> 56));
    }

    public void WriteFloat(float fvalue)
    {
        long value = BitConverter.SingleToInt32Bits(fvalue);
        memoryStream.WriteByte((byte)value);
        memoryStream.WriteByte((byte)(value >> 8));
        memoryStream.WriteByte((byte)(value >> 16));
        memoryStream.WriteByte((byte)(value >> 24));
    }

    public void WriteDouble(double dvalue)
    {
        long value = BitConverter.DoubleToInt64Bits(dvalue);
        memoryStream.WriteByte((byte)value);
        memoryStream.WriteByte((byte)(value >> 8));
        memoryStream.WriteByte((byte)(value >> 16));
        memoryStream.WriteByte((byte)(value >> 24));
        memoryStream.WriteByte((byte)(value >> 32));
        memoryStream.WriteByte((byte)(value >> 40));
        memoryStream.WriteByte((byte)(value >> 48));
        memoryStream.WriteByte((byte)(value >> 56));
    }

    public byte[] GetSendableBytes()
    {
        // Encrypt bytes.
        byte[] encryptedBytes = Encryption.Encrypt(memoryStream.ToArray());
        int size = encryptedBytes.Length;

        // Create two bytes for length (short - max length 32767).
        byte[] lengthBytes = new byte[2];
        lengthBytes[0] = (byte)(size & 0xff);
        lengthBytes[1] = (byte)((size >> 8) & 0xff);

        // Join bytes.
        byte[] result = new byte[size + 2];
        Array.Copy(lengthBytes, 0, result, 0, 2);
        Array.Copy(encryptedBytes, 0, result, 2, size);

        // Return the data.
        return result;
    }
}
